File: integration-tests/c/sample-rejected-lost/src/publisher.c
/*
 * integration-tests/c/sample-rejected-lost -- publisher.
 *
 * Three DataWriters under one Publisher, written in this specific order:
 * LostTopic (RELIABLE, KEEP_LAST depth=1, TRANSIENT_LOCAL) gets 5
 * back-to-back writes FIRST, before any reader can possibly be matched to
 * it -- the subscriber deliberately defers creating that reader until
 * after the Sync signal below, so these writes are written and evicted
 * with zero chance of ever being acked (a genuinely, deterministically
 * un-recoverable loss for SNs 1-4, not a race against real-time ack
 * latency -- an earlier version of this scenario tried racing fast writes
 * against an already-matched reader and nothing was ever lost, because
 * localhost round-trips are fast enough that each sample got acked before
 * the next write evicted it). Then RejectedTopic (RELIABLE, KEEP_ALL, no
 * resource limits of its own -- the *reader's* tight resource_limits is
 * what causes rejection) gets 5 back-to-back writes against an
 * already-matched reader. SyncTopic gets one write *after* everything
 * above is done, purely so the subscriber knows it's safe to create the
 * LostTopic reader and check final status without racing the publisher's
 * own writes. See docs/design/integration-test-tier.md for the full
 * scenario spec.
 *
 * Required stdout markers: "Create topic:" x3, "Create writer for topic:"
 * x3, "Publisher: wrote 5 samples on LostTopic...", "Publisher: wrote 5
 * samples on RejectedTopic...", "Publisher: sync sent.", "Publisher: done."
 * Any failure path prints a line starting "FAIL:" and exits nonzero.
 */
#include "status_event.h"
#include "zzdds_c.h"
#include "zzdds.h"

#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define READER_READY_TIMEOUT_MS 10000
#define DRAIN_TIMEOUT_MS 15000
#define POLL_PERIOD_MS 20

typedef struct {
    atomic_bool reader_ready;
    atomic_bool ever_matched;
    atomic_int matched_current_count;
} WriterSyncState;

static void on_reliable_reader_ready(DDS_InstanceHandle_t reader_handle, bool is_ready, void *listener_data) {
    (void)reader_handle;
    WriterSyncState *state = (WriterSyncState *)listener_data;
    if (is_ready) atomic_store(&state->reader_ready, true);
}

static void on_publication_matched(DDS_DataWriter writer, const DDS_PublicationMatchedStatus *status, void *listener_data) {
    (void)writer;
    WriterSyncState *state = (WriterSyncState *)listener_data;
    atomic_store(&state->matched_current_count, status->current_count);
    if (status->current_count > 0) atomic_store(&state->ever_matched, true);
}

static uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (strcmp(argv[i], "-d") == 0 || strcmp(argv[i], "--domain") == 0) {
            return (uint32_t)strtoul(argv[i + 1], NULL, 10);
        }
    }
    return 0;
}

static int set_writer_listener(DDS_DataWriter dw, WriterSyncState *state) {
    zzdds_DataWriter zdw = DDS_DataWriter_as_zzdds_DataWriter(dw);
    zzdds_DataWriterListenerEx listener_ex;
    memset(&listener_ex, 0, sizeof(listener_ex));
    listener_ex.listener_data = state;
    listener_ex.on_publication_matched = on_publication_matched;
    listener_ex.on_reliable_reader_ready = on_reliable_reader_ready;
    return zzdds_DataWriter_set_listener_ex(zdw, &listener_ex, DDS_PUBLICATION_MATCHED_STATUS);
}

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    zzdds_DomainParticipantFactory factory = zzdds_create_factory();
    if (zzdds_factory_is_nil(factory)) {
        fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }
    DDS_DomainParticipantFactory dds_factory = zzdds_DomainParticipantFactory_as_DDS_DomainParticipantFactory(factory);

    DDS_DomainParticipant dp = DDS_DomainParticipantFactory_create_participant(dds_factory, domain_id, NULL, NULL, 0);
    if (!dp) {
        fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }

    if (StatusEventTypeSupport_register(dp, "StatusEvent") != DDS_RETCODE_OK) {
        fprintf(stderr, "FAIL: register StatusEventTypeSupport failed\n");
        return 1;
    }

    DDS_Topic rejected_topic = DDS_DomainParticipant_create_topic(dp, "RejectedTopic", "StatusEvent", NULL, NULL, 0);
    if (!rejected_topic) {
        fprintf(stderr, "FAIL: create_topic(RejectedTopic) failed\n");
        return 1;
    }
    printf("Create topic: RejectedTopic\n");

    DDS_Topic lost_topic = DDS_DomainParticipant_create_topic(dp, "LostTopic", "StatusEvent", NULL, NULL, 0);
    if (!lost_topic) {
        fprintf(stderr, "FAIL: create_topic(LostTopic) failed\n");
        return 1;
    }
    printf("Create topic: LostTopic\n");

    DDS_Topic sync_topic = DDS_DomainParticipant_create_topic(dp, "SyncTopic", "StatusEvent", NULL, NULL, 0);
    if (!sync_topic) {
        fprintf(stderr, "FAIL: create_topic(SyncTopic) failed\n");
        return 1;
    }
    printf("Create topic: SyncTopic\n");

    DDS_Publisher pub = DDS_DomainParticipant_create_publisher(dp, NULL, NULL, 0);
    if (!pub) {
        fprintf(stderr, "FAIL: create_publisher() failed\n");
        return 1;
    }

    /* LostTopic FIRST, deliberately before any reader can possibly be
     * matched -- see the file header comment. */
    DDS_DataWriterQos lost_qos;
    DDS_Publisher_get_default_datawriter_qos(pub, &lost_qos);
    lost_qos.reliability.kind = DDS_ReliabilityQosPolicyKind_RELIABLE_RELIABILITY_QOS;
    lost_qos.history.kind = DDS_HistoryQosPolicyKind_KEEP_LAST_HISTORY_QOS;
    lost_qos.history.depth = 1;
    /* TRANSIENT_LOCAL so the late-joining reader can still receive whatever
     * remains in the writer's cache (the seq=4 sample) -- VOLATILE (the
     * default) would make it miss ALL 5 samples, not just the 4 evicted
     * ones. Eviction (and therefore loss of seq 0-3) still happens
     * regardless of durability. */
    lost_qos.durability.kind = DDS_DurabilityQosPolicyKind_TRANSIENT_LOCAL_DURABILITY_QOS;

    DDS_DataWriter lost_dw = DDS_Publisher_create_datawriter(pub, lost_topic, &lost_qos, NULL, 0);
    if (!lost_dw) {
        fprintf(stderr, "FAIL: create_datawriter(LostTopic) failed\n");
        return 1;
    }
    printf("Create writer for topic: LostTopic\n");

    StatusEventDataWriter lost_writer;
    StatusEventDataWriter_init(&lost_writer, lost_dw, ZIDL_XCDR1);
    for (int seq = 0; seq < 5; seq++) {
        StatusEvent ev;
        memset(&ev, 0, sizeof(ev));
        ev.seq = seq;
        if (StatusEventDataWriter_write(&lost_writer, &ev, DDS_HANDLE_NIL) != 0) {
            fprintf(stderr, "FAIL: LostTopic write() failed at seq=%d\n", seq);
            return 1;
        }
    }
    printf("Publisher: wrote 5 samples on LostTopic (KEEP_LAST depth=1, no reader matched yet).\n");

    DDS_DataWriterQos rejected_qos;
    DDS_Publisher_get_default_datawriter_qos(pub, &rejected_qos);
    rejected_qos.reliability.kind = DDS_ReliabilityQosPolicyKind_RELIABLE_RELIABILITY_QOS;
    rejected_qos.history.kind = DDS_HistoryQosPolicyKind_KEEP_ALL_HISTORY_QOS;
    DDS_DataWriter rejected_dw = DDS_Publisher_create_datawriter(pub, rejected_topic, &rejected_qos, NULL, 0);
    if (!rejected_dw) {
        fprintf(stderr, "FAIL: create_datawriter(RejectedTopic) failed\n");
        return 1;
    }
    printf("Create writer for topic: RejectedTopic\n");

    DDS_DataWriterQos sync_qos;
    DDS_Publisher_get_default_datawriter_qos(pub, &sync_qos);
    sync_qos.reliability.kind = DDS_ReliabilityQosPolicyKind_RELIABLE_RELIABILITY_QOS;
    DDS_DataWriter sync_dw = DDS_Publisher_create_datawriter(pub, sync_topic, &sync_qos, NULL, 0);
    if (!sync_dw) {
        fprintf(stderr, "FAIL: create_datawriter(SyncTopic) failed\n");
        return 1;
    }
    printf("Create writer for topic: SyncTopic\n");

    WriterSyncState rejected_state, sync_state, lost_state;
    memset(&rejected_state, 0, sizeof(rejected_state));
    memset(&sync_state, 0, sizeof(sync_state));
    memset(&lost_state, 0, sizeof(lost_state));

    if (set_writer_listener(rejected_dw, &rejected_state) != DDS_RETCODE_OK ||
        set_writer_listener(sync_dw, &sync_state) != DDS_RETCODE_OK ||
        set_writer_listener(lost_dw, &lost_state) != DDS_RETCODE_OK)
    {
        fprintf(stderr, "FAIL: set_listener_ex failed\n");
        return 1;
    }

    /* Only Rejected/Sync need to wait for their reader -- the subscriber
     * creates those two immediately at startup. LostTopic's reader isn't
     * created until the subscriber gets the Sync sample, by design. */
    for (int waited_ms = 0;
         !(atomic_load(&rejected_state.reader_ready) && atomic_load(&sync_state.reader_ready));
         waited_ms += POLL_PERIOD_MS)
    {
        if (waited_ms >= READER_READY_TIMEOUT_MS) {
            fprintf(stderr, "FAIL: no reliable reader became ready within %ds\n", READER_READY_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    StatusEventDataWriter rejected_writer;
    StatusEventDataWriter_init(&rejected_writer, rejected_dw, ZIDL_XCDR1);
    /* The subscriber deliberately does not drain RejectedTopic until it has
     * confirmed rejection happened, so no reader-side consumption race is
     * possible here regardless of exact write timing. */
    for (int seq = 0; seq < 5; seq++) {
        StatusEvent ev;
        memset(&ev, 0, sizeof(ev));
        ev.seq = seq;
        if (StatusEventDataWriter_write(&rejected_writer, &ev, DDS_HANDLE_NIL) != 0) {
            fprintf(stderr, "FAIL: RejectedTopic write() failed at seq=%d\n", seq);
            return 1;
        }
    }
    printf("Publisher: wrote 5 samples on RejectedTopic (up to 2 expected rejected).\n");

    StatusEventDataWriter sync_writer;
    StatusEventDataWriter_init(&sync_writer, sync_dw, ZIDL_XCDR1);
    StatusEvent sync_ev;
    memset(&sync_ev, 0, sizeof(sync_ev));
    if (StatusEventDataWriter_write(&sync_writer, &sync_ev, DDS_HANDLE_NIL) != 0) {
        fprintf(stderr, "FAIL: SyncTopic write() failed\n");
        return 1;
    }
    printf("Publisher: sync sent.\n");
    printf("Publisher: done.\n");

    for (int waited_ms = 0;
         !(atomic_load(&rejected_state.ever_matched) && atomic_load(&rejected_state.matched_current_count) == 0 &&
           atomic_load(&sync_state.ever_matched) && atomic_load(&sync_state.matched_current_count) == 0 &&
           atomic_load(&lost_state.ever_matched) && atomic_load(&lost_state.matched_current_count) == 0);
         waited_ms += POLL_PERIOD_MS)
    {
        if (waited_ms >= DRAIN_TIMEOUT_MS) {
            fprintf(stderr, "FAIL: subscriber did not disconnect within %ds\n", DRAIN_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    if (DDS_Publisher_delete_datawriter(pub, rejected_dw) != DDS_RETCODE_OK ||
        DDS_Publisher_delete_datawriter(pub, lost_dw) != DDS_RETCODE_OK ||
        DDS_Publisher_delete_datawriter(pub, sync_dw) != DDS_RETCODE_OK)
    {
        fprintf(stderr, "FAIL: delete_datawriter() did not return RETCODE_OK\n");
        return 1;
    }

    zzdds_destroy_factory(factory);
    return 0;
}

File: integration-tests/zig/sample-rejected-lost/publisher.zig
//! zig/sample-rejected-lost -- publisher.
//!
//! Three DataWriters under one Publisher, written in this specific order:
//! LostTopic (RELIABLE, KEEP_LAST depth=1, TRANSIENT_LOCAL) gets 5
//! back-to-back writes FIRST, before any reader can possibly be matched to
//! it -- the subscriber deliberately defers creating that reader until
//! after the Sync signal below, so these writes are written and evicted
//! with zero chance of ever being acked (a genuinely, deterministically
//! un-recoverable loss for SNs 1-4, not a race against real-time ack
//! latency -- an earlier version of this scenario tried racing fast writes
//! against an already-matched reader and nothing was ever lost, because
//! localhost round-trips are fast enough that each sample got acked before
//! the next write evicted it). Then RejectedTopic (RELIABLE, KEEP_ALL, no
//! resource limits of its own -- the *reader's* tight resource_limits is
//! what causes rejection) gets 5 back-to-back writes against an
//! already-matched reader. SyncTopic gets one write *after* everything
//! above is done, purely so the subscriber knows it's safe to create the
//! LostTopic reader and check final status without racing the publisher's
//! own writes. See docs/design/integration-test-tier.md for the full
//! scenario spec.
//!
//! Required stdout markers: "Create topic:" x3, "Create writer for topic:"
//! x3, "Publisher: wrote 5 samples on LostTopic...", "Publisher: wrote 5
//! samples on RejectedTopic...", "Publisher: sync sent.", "Publisher: done."
//! Any failure path prints a line starting "FAIL:" and exits nonzero.

const std = @import("std");
const zzdds = @import("zzdds");
const DDS = @import("zzdds_generated").DDS;
const ZZDDS = @import("zzdds_ext_generated").zzdds;
const status_event_gen = @import("status_event_gen");

const READER_READY_TIMEOUT_NS: i64 = 10 * std.time.ns_per_s;
const DRAIN_TIMEOUT_NS: i64 = 15 * std.time.ns_per_s;
const POLL_PERIOD_NS: u64 = 20 * std.time.ns_per_ms;

fn monoNs(io: std.Io) i64 {
    return @intCast(std.Io.Clock.awake.now(io).nanoseconds);
}

fn sleepNs(io: std.Io, ns: u64) void {
    (std.Io.Clock.Duration{ .raw = .{ .nanoseconds = @intCast(ns) }, .clock = .awake }).sleep(io) catch {};
}

const WriterSyncState = struct {
    reader_ready: std.atomic.Value(bool) = std.atomic.Value(bool).init(false),
    matched_current_count: std.atomic.Value(i32) = std.atomic.Value(i32).init(0),
    ever_matched: std.atomic.Value(bool) = std.atomic.Value(bool).init(false),
};

fn onReliableReaderReady(state: *WriterSyncState, reader_handle: DDS.InstanceHandle_t, is_ready: bool) void {
    _ = reader_handle;
    if (is_ready) state.reader_ready.store(true, .release);
}

fn onPublicationMatched(state: *WriterSyncState, dw: DDS.DataWriter, status: DDS.PublicationMatchedStatus) void {
    _ = dw;
    state.matched_current_count.store(status.current_count, .release);
    if (status.current_count > 0) state.ever_matched.store(true, .release);
}

fn parseDomain(process_args: std.process.Args) u32 {
    var it = std.process.Args.Iterator.init(process_args);
    _ = it.skip();
    while (it.next()) |arg| {
        if (std.mem.eql(u8, arg, "-d") or std.mem.eql(u8, arg, "--domain")) {
            const v = it.next() orelse continue;
            return std.fmt.parseInt(u32, v, 10) catch 0;
        }
    }
    return 0;
}

fn setWriterListener(dw: DDS.DataWriter, state: *WriterSyncState) !void {
    const zdw = zzdds.asZzddsDataWriter(dw) orelse return error.AsZzddsDataWriterFailed;
    if (zdw.set_listener_ex(ZZDDS.dataWriterListenerEx(state, .{
        .on_publication_matched = onPublicationMatched,
        .on_reliable_reader_ready = onReliableReaderReady,
    }), DDS.PUBLICATION_MATCHED_STATUS) != DDS.RETCODE_OK) return error.SetListenerExFailed;
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    var gpa = std.heap.DebugAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    const domain_id = parseDomain(init.minimal.args);

    var factory = zzdds.createFactory() catch {
        std.debug.print("FAIL: createFactory() failed\n", .{});
        std.process.exit(1);
    };
    defer factory.deinit();
    const dpf = factory.toDDSFactory();

    const dp = dpf.create_participant(domain_id, .{}, null, 0);
    if (dp.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_participant() failed on domain {d}\n", .{domain_id});
        std.process.exit(1);
    }
    defer _ = dpf.delete_participant(dp);

    var ts_alloc = alloc;
    if (!zzdds.registerTypeSupport(dp, "StatusEvent", .{
        .ctx = @ptrCast(&ts_alloc),
        .compute_key_hash = status_event_gen.StatusEvent.computeKeyHashFromCdr,
        .compute_key_hash_key_only = status_event_gen.StatusEvent.computeKeyHashFromCdrKeyOnly,
    })) {
        std.debug.print("FAIL: registerTypeSupport(StatusEvent) failed\n", .{});
        std.process.exit(1);
    }

    const topics = [_][:0]const u8{ "RejectedTopic", "LostTopic", "SyncTopic" };
    var topic_descs: [3]DDS.Topic = undefined;
    for (topics, 0..) |name, i| {
        const t = dp.create_topic(name, "StatusEvent", .{}, null, 0);
        if (t.ptr == zzdds.dcps.NIL_PTR) {
            std.debug.print("FAIL: create_topic({s}) failed\n", .{name});
            std.process.exit(1);
        }
        std.debug.print("Create topic: {s}\n", .{name});
        topic_descs[i] = t;
    }

    const publisher = dp.create_publisher(.{}, null, 0);
    if (publisher.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_publisher() failed\n", .{});
        std.process.exit(1);
    }

    // LostTopic FIRST, deliberately before any reader can possibly be
    // matched: the subscriber defers creating its LostTopic reader until
    // after the Sync signal below, specifically so these 5 writes (KEEP_LAST
    // depth=1) are written and evicted with zero chance of ever being acked
    // -- a genuinely, deterministically un-recoverable loss, not a race
    // against real-time ack latency (which turned out on localhost to
    // always win: an empirical first attempt at this scenario wrote+evicted
    // fast against an ALREADY-matched reader and nothing was ever lost,
    // because each sample got acked before the next write evicted it -- see
    // docs/design/integration-test-tier.md's scenario notes).
    var lost_qos = DDS.DataWriterQos{};
    lost_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    lost_qos.history.kind = .KEEP_LAST_HISTORY_QOS;
    lost_qos.history.depth = 1;
    // TRANSIENT_LOCAL so the late-joining reader can still receive whatever
    // remains in the writer's cache (the seq=4 sample) -- VOLATILE (the
    // default) would make it miss ALL 5 samples, not just the 4 evicted
    // ones, since VOLATILE's contract is "matched before write time or
    // nothing," independent of what the writer still happens to have
    // cached. Eviction (and therefore loss of seq 0-3) still happens
    // regardless of durability -- TRANSIENT_LOCAL only governs whether a
    // late joiner can reach what's currently cached, not whether KEEP_LAST
    // still evicts.
    lost_qos.durability.kind = .TRANSIENT_LOCAL_DURABILITY_QOS;
    const lost_dw = publisher.create_datawriter(topic_descs[1], lost_qos, null, 0);
    if (lost_dw.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_datawriter(LostTopic) failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create writer for topic: LostTopic\n", .{});

    {
        const lost_writer = status_event_gen.StatusEventDataWriter.init(lost_dw, alloc);
        var seq: i32 = 0;
        while (seq < 5) : (seq += 1) {
            lost_writer.write(.{ .seq = seq }, 0) catch {
                std.debug.print("FAIL: LostTopic write() failed at seq={d}\n", .{seq});
                std.process.exit(1);
            };
        }
    }
    std.debug.print("Publisher: wrote 5 samples on LostTopic (KEEP_LAST depth=1, no reader matched yet).\n", .{});

    var rejected_qos = DDS.DataWriterQos{};
    rejected_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    rejected_qos.history.kind = .KEEP_ALL_HISTORY_QOS;
    const rejected_dw = publisher.create_datawriter(topic_descs[0], rejected_qos, null, 0);
    if (rejected_dw.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_datawriter(RejectedTopic) failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create writer for topic: RejectedTopic\n", .{});

    var sync_qos = DDS.DataWriterQos{};
    sync_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    const sync_dw = publisher.create_datawriter(topic_descs[2], sync_qos, null, 0);
    if (sync_dw.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_datawriter(SyncTopic) failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create writer for topic: SyncTopic\n", .{});

    var rejected_state = WriterSyncState{};
    var sync_state = WriterSyncState{};
    var lost_state = WriterSyncState{};
    setWriterListener(rejected_dw, &rejected_state) catch {
        std.debug.print("FAIL: set_listener_ex(RejectedTopic) failed\n", .{});
        std.process.exit(1);
    };
    setWriterListener(sync_dw, &sync_state) catch {
        std.debug.print("FAIL: set_listener_ex(SyncTopic) failed\n", .{});
        std.process.exit(1);
    };
    // Only used for the final drain-wait below -- LostTopic's reader isn't
    // created until well after this writer already exists, so there's
    // nothing to gate the first write on here.
    setWriterListener(lost_dw, &lost_state) catch {
        std.debug.print("FAIL: set_listener_ex(LostTopic) failed\n", .{});
        std.process.exit(1);
    };

    // Only Rejected/Sync need to wait for their reader -- the subscriber
    // creates those two immediately at startup. LostTopic's reader isn't
    // created until the subscriber gets the Sync sample below, by design.
    const ready_deadline = monoNs(io) + READER_READY_TIMEOUT_NS;
    while (!(rejected_state.reader_ready.load(.acquire) and sync_state.reader_ready.load(.acquire))) {
        if (monoNs(io) > ready_deadline) {
            std.debug.print("FAIL: no reliable reader became ready within 10s\n", .{});
            std.process.exit(1);
        }
        sleepNs(io, POLL_PERIOD_NS);
    }

    const rejected_writer = status_event_gen.StatusEventDataWriter.init(rejected_dw, alloc);
    const sync_writer = status_event_gen.StatusEventDataWriter.init(sync_dw, alloc);

    // The subscriber deliberately does not drain RejectedTopic until it has
    // confirmed rejection happened, so no reader-side consumption race is
    // possible here regardless of exact write timing.
    var seq: i32 = 0;
    while (seq < 5) : (seq += 1) {
        rejected_writer.write(.{ .seq = seq }, 0) catch {
            std.debug.print("FAIL: RejectedTopic write() failed at seq={d}\n", .{seq});
            std.process.exit(1);
        };
    }
    std.debug.print("Publisher: wrote 5 samples on RejectedTopic (up to 2 expected rejected).\n", .{});

    sync_writer.write(.{ .seq = 0 }, 0) catch {
        std.debug.print("FAIL: SyncTopic write() failed\n", .{});
        std.process.exit(1);
    };
    std.debug.print("Publisher: sync sent.\n", .{});
    std.debug.print("Publisher: done.\n", .{});

    const drain_deadline = monoNs(io) + DRAIN_TIMEOUT_NS;
    while (!(rejected_state.ever_matched.load(.acquire) and rejected_state.matched_current_count.load(.acquire) == 0 and
        lost_state.ever_matched.load(.acquire) and lost_state.matched_current_count.load(.acquire) == 0 and
        sync_state.ever_matched.load(.acquire) and sync_state.matched_current_count.load(.acquire) == 0))
    {
        if (monoNs(io) > drain_deadline) {
            std.debug.print("FAIL: subscriber did not disconnect within 15s\n", .{});
            std.process.exit(1);
        }
        sleepNs(io, POLL_PERIOD_NS);
    }

    if (publisher.delete_datawriter(rejected_dw) != DDS.RETCODE_OK or
        publisher.delete_datawriter(lost_dw) != DDS.RETCODE_OK or
        publisher.delete_datawriter(sync_dw) != DDS.RETCODE_OK)
    {
        std.debug.print("FAIL: delete_datawriter() did not return RETCODE_OK\n", .{});
        std.process.exit(1);
    }
}
